executor/read/pattern_executor.rs

```diff
@@ -382,26 +478,32 @@ impl PatternExecutor {
         let found = match executor.try_read_next_batch(&function_state) {
             TabledCallResult::RetrievedFromTable(batch) => Some(batch),
             TabledCallResult::Suspend => {
-                suspend_point_accumulator.push(executor.create_suspend_point_for(index));
+                suspend_point_context.push_tabled_call(index, executor); // Clearly a dummy
                 None
             }
             TabledCallResult::MustExecutePattern(mut pattern_state_mutex_guard) => {
-                let TabledFunctionPatternExecutorState { pattern_executor, suspend_points, parameters } =
-                    pattern_state_mutex_guard.deref_mut();
+                let TabledFunctionPatternExecutorState {
+                    pattern_executor,
+                    suspend_points: function_suspend_points,
+                    parameters,
+                } = pattern_state_mutex_guard.deref_mut();
                 let context_with_function_parameters =
                     ExecutionContext::new(context.snapshot.clone(), context.thing_manager.clone(), parameters.clone());
+                let _ignore_suspension_state_before = function_suspend_points.record_nested_pattern_entry();
                 let batch_opt = pattern_executor.batch_continue(
                     &context_with_function_parameters,
                     interrupt,
                     tabled_functions,
-                    suspend_points,
+                    function_suspend_points,
                 )?;
+                let suspension_state_after = function_suspend_points.record_nested_pattern_exit();
                 if let Some(batch) = batch_opt {
                     let deduplicated_batch = executor.add_batch_to_table(&function_state, batch);
                     Some(deduplicated_batch)
                 } else {
-                    if !suspend_points.is_empty() {
-                        suspend_point_accumulator.push(executor.create_suspend_point_for(index))
+                    if suspension_state_after.has_suspend_points() {
+                        // TODO: Consider retrying here. For now, just record a suspend point for ourselves.
+                        suspend_point_context.push_tabled_call(index, executor);
```

Comment:
So if i'm getting this right - every function call & the entry point has a set of suspend points (loosely, which parts of the query that is being executed were signalled to pause. If a function hits suspend points, we also record a suspend point in the parent caller.

Reply:
Yes. Because we need to restore the whole stack till that point.